### PR TITLE
feat(core): add custom headers capability for application response and make `JWTAuth` private

### LINKED ISSUE #1291: [ENHANCEMENT] - Add custom headers support while sending response
### Feature Description

Currently, there is no way to control the headers that are being sent in the response. Adding support for that would be a nice feature that many other features could benefit from.

### Possible Implementation

Create a new `ApplicationResponse` which can take custom headers and attach them to the `HttpResponse`.

### Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ThisIsMani**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Add custom headers capability to application response. 
- As we are not using `JWTAuth` in other crates., this one is made private.
- `jwt_auth_or` function is removed, which can be replaced by `auth_type`.
- Create a new `ApiErrorResponse` for 403 (`AccessForbidden`).

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

As we can only add headers to `HttpResponse`, which can be only used in `server_wrap`, it's better if we can control it in the inner function itself, which only has access to `ApplicationResponse`. So I have added this new response type.
Closes #1291 

## How did you test it?

Postman

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/compatibility/wrap.rs [hunk 1]
```diff
@@ -40,7 +40,28 @@ where
...
+) -> HttpResponse
+where
+    Q: Serialize + std::fmt::Debug,
+    E: Serialize + error_stack::Context + actix_web::ResponseError + Clone,
+    S: TryFrom<Q> + Serialize,
+    error_stack::Report<E>: services::EmbedError,
+{
+    match server_resp {
+        Ok(api::ApplicationResponse::ResponseWithCustomHeader(response, headers)) => {
+            let mut final_response = handle_application_response::<Q, S, E>(request, Ok(*response));
```
Comment (jarnura) [inline crates/router/src/compatibility/wrap.rs:58]: Why the function calling itself?

**Review by jarnura — CHANGES_REQUESTED** [review]
Add sub-issue, milestone, projects.

File: crates/router/src/compatibility/wrap.rs (hunk 1 above)
Comment (ThisIsMani) [reply crates/router/src/compatibility/wrap.rs:58]: `ResponseWithCustomHeader` has two fields. The first field is `response` of type `ApplicationResponse`, and the second field is `headers` which is a `HeaderMap`. The function `handle_application_response` takes an `ApplicationResponse` as input and returns an `HttpResponse`. To obtain the `HttpResponse` for the inner `response`, the function is recursively called again. Similar behavior can be observed in other functions as well.

**Comment by lsampras** [comment]
@ThisIsMani 
As discussed offline,
can we try to do this via an actix middleware,

you can close this PR if that approach works out...

cc: @SanchithHegde @jarnura

**Comment by ThisIsMani** [comment]
This is closed. Actix middleware and server_wrap is being used to achieve this.

**Pull request closed without merging**